### DISCUSSION TITLE
Loading screen widget?

### DISCUSSION THREAD

**Discussion by youstanzr** [israel-dryer/ttkbootstrap#276] [Q&A]
I am building an app that generates some files and the processing takes a long time to complete. Throughout the time its generating, I was hoping I can overlay a frame layer on top of the app screen to prevent the user to use any widgets on the screen and show the loading state.

I was wondering if ttkbootstrap has a widget to simply show a loading screen? if not, whats the best way to go about it with the current available widgets?

Thanks,

**Comment by israel-dryer** [comment]
There are a lot of ways to do this. They key is to make sure you are running your file processing on a separate thread to keep the IO operations from blocking the GUI event loop. This will cause the window to become unresponsive.

You could use some kind of [loading screen with a GIF](https://ttkbootstrap.readthedocs.io/en/latest/cookbook/gif-animation/).

Or, you could `pack_forget` the main frame and add another frame into the window while the process is taking place, the pack the original frame back in. This works if you want to use the same window instead of using multiple windows.

Or, you could use a traditional Toplevel window with message and progress bar like the example below. I've added comments to explain various components.

To keep the user from interacting with the screen, you could either disable the widgets, or you could use a simplistic approach like I have below and use the `grab_set` method, which essentially routes all application events to the Toplevel widget. This means that nothing the user does on the main window will have any effect until the grab is released.

![Code_PCe2hGEMzh](https://user-images.githubusercontent.com/42584742/164977582-3c6c4a3f-1226-41a6-9dcb-48d4d0b21c33.gif)


```python
import ttkbootstrap as ttk
from ttkbootstrap.constants import *
from threading import Thread
from time import sleep

class Application(ttk.Window):

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        # hide window until widgets are drawn
        self.withdraw()

        # add a few widgets to the main window
        lbl_text = """Click the button to process all of the files"""
        ttk.Label(self, text=lbl_text).pack(padx=20, pady=20)
        btn = ttk.Button(self, text='Process', command=self.process_files)
        btn.pack(pady=20)

        # center the window and show
        self.position_center() 
        self.deiconify()       
        
    def process_files(self):
        # Create a loading screen with progress bar
        ls = LoadingScreen(transient=self, title="Loading files...")
        
        # Create a process handler that will be run on a separate thread
        lr = LongRunningProcess(ls.msg_var, ls.is_running)
        t = Thread(daemon=True, target=lr.run)
        t.start()


class LoadingScreen(ttk.Toplevel):

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.position_center()
        
        # grab all window events to keep user from using another screen
        self.grab_current()

        # add variable with trace to keep track of loading state        
        self.is_running = ttk.BooleanVar(self, True)
        self.is_running.trace_add("write", self.shutdown)

        # add message variable that can be updated by a threaded process
        self.msg_var = ttk.StringVar(self, "Working...please wait...")
        self.message = ttk.Label(self, textvariable=self.msg_var)
        self.message.pack()
        
        # indeterminate progressbar can be used when total tasks are unknown
        self.progress = ttk.Progressbar(self, mode=INDETERMINATE)
        self.progress.pack(fill=X, padx=20, pady=20)
        self.progress.start()

    def shutdown(self, *args):
        # give control back to main window and destroy this widget
        self.grab_release()
        self.destroy()    

class LongRunningProcess:

    def __init__(self, message_var, is_running):
        self.message_var = message_var
        self.is_running = is_running
        
    def run(self):
        # add the code for your long-running process here
        for x in range(1, 6):
            # update the message while you completing the work.
            self.message_var.set(f"Processing file {x}.")
            sleep(1) # simulates blocking IO operations 
        
        # tell the loading screen that you are done
        self.is_running.set(False)

if __name__ == '__main__':

    app = Application()
    app.mainloop()
```

**Reply by youstanzr** [reply]
What if the user just closed the loading screen? the thread will keep running but no loading screen appears

**Reply by israel-dryer** [reply]
> What if the user just closed the loading screen? the thread will keep running but no loading screen appears

Yes, the thread will still run as long as the main application window is alive even if you close the loading screen. However, the thread will die if you close the program entirely as long as it's opened as a daemon thread.

What behavior do you want? Do you want to prevent the user from closing the loading screen?

**Reply by youstanzr** [reply]
It would be best to kill the thread if the user closed the loading screen since that counts as canceling the operation

**Reply by israel-dryer** [reply]
I'm not sure the best or [safest way to kill the thread in python](https://www.geeksforgeeks.org/python-different-ways-to-kill-a-thread/), but if you decide on a method, the way to bind it to the loading screen is to update the `WM_DELETE_WINDOW` protocol on the toplevel.

within the loading screen `__init__` method in the example above... add the following...

```python
# method example without arguments
self.protocol("WM_DELETE_WINDOW", self.your_kill_thread_method)

# method example with arguments
self.protocol("WM_DELETE_WINDOW", lambda x=1, y=2, z=3: self.your_kill_thread_method(x, y, z))
```

When you close the loading window, this method will get called.

**Reply by TheDemonGuard** [reply]
@youstanzr 

It is possible to set `overrideredirect(True)` to remove all window manager decorations from the splash screen window (Toplevel),
so that it cannot be closed.

| Before | After |
|---------|-------|
| <img width="331" height="240" alt="splashScreen0" src="https://github.com/user-attachments/assets/707d14b1-2c69-4e80-a8c5-853fed04c6f0" /> | <img width="327" height="231" alt="splashScreen1" src="https://github.com/user-attachments/assets/b6334a79-e9d1-45fc-9f43-fdc93d66fbfb" /> |

That been said, `ttkbootstrap` is a library to offer GUI widgets.
Therefore, backend and/or underlying operation(s) should be implemented by the application developer.
Hence, it is your responsibility to implement how to safely kill the thread and close the application gracefully.

> It would be best to kill the thread if the user closed the loading screen since that counts as canceling the operation
